Dockerfile

```diff
@@ -39,65 +43,50 @@ RUN apt-get update \
     nodejs \
     npm \
  && apt-get clean \
- && rm -rf /var/lib/apt/lists/*
-
-RUN python3 -m pip install --upgrade setuptools pip build wheel
-RUN npm install --global yarn
-
-# copy everything except whats in .dockerignore, its a
-# compromise between needing to rebuild and maintaining
-# what needs to be part of the build
-COPY . /src/jupyterhub/
-WORKDIR /src/jupyterhub
+ && rm -rf /var/lib/apt/lists/* \
+ && python3 -m pip install --no-cache-dir --upgrade setuptools pip build wheel \
+ && npm install --global yarn \
+ # Build client component packages (they will be copied into ./share and
+ # packaged with the built wheel.)
+ && python3 -m build --wheel \
```

Comment:
I will note that reducing layer count in the build stage only has a downside in build caching. It shouldn't affect the size of the resulting image. I don't think this change is an improvement.